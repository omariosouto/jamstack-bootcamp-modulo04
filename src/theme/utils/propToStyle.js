import { breakpointsMedia } from './breakpointsMedia';

export function propToStyle(propName) {
  return (props) => {
    const propValue = props[propName];

    if (typeof propValue === 'string' || typeof propValue === 'number') {
      return {
        // textAlign: props.textAlign
        [propName]: propValue,
      };
    }

    if (typeof propValue === 'object') {
      const breakpoints = {};
      if (propValue.xs) breakpoints.xs = { [propName]: propValue.xs };
      if (propValue.sm) breakpoints.sm = { [propName]: propValue.sm };
      if (propValue.md) breakpoints.md = { [propName]: propValue.md };
      if (propValue.lg) breakpoints.lg = { [propName]: propValue.lg };
      if (propValue.xl) breakpoints.xl = { [propName]: propValue.xl };

      return breakpointsMedia(breakpoints);
    }

    return {};
  };
}
