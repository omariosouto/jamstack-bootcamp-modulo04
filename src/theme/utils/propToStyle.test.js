import { propToStyle } from './propToStyle';

describe('propToStyle()', () => {
  describe('when receives a sample argument', () => {
    test('and it is a string', () => {
      const propToStyleResult = propToStyle('color');
      expect(typeof propToStyleResult).toBe('function');

      // <Text color='red' />
      const componentProps = { color: 'red' };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toEqual({ color: 'red' });
    });
    test('and it is a number', () => {
      const propToStyleResult = propToStyle('flex');
      expect(typeof propToStyleResult).toBe('function');

      // <Grid.Col flex={1} />
      const componentProps = { flex: 1 };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toEqual({ flex: 1 });
    });
  });

  describe('when receive an argument with breakpoints', () => {
    test('renders only the received ones', () => {
      const propToStyleResult = propToStyle('color');
      // <Text color={{ xs: 'red', md: 'blue' }} />
      const componentProps = {
        color: {
          xs: 'red',
          md: 'blue',
        },
      };
      const styleResult = propToStyleResult(componentProps);
      expect(styleResult).toMatchSnapshot();
    });
  });
});
