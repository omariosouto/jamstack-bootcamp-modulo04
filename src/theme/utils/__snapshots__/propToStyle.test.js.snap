// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propToStyle() when receive an argument with breakpoints renders only the received ones 1`] = `
Array [
  Array [
    "@media screen and (min-width:",
    "0",
    "px){",
    "color: red;",
    "}",
  ],
  Array [
    "@media screen and (min-width:",
    "768",
    "px){",
    "color: blue;",
    "}",
  ],
]
`;
